Current architecture snapshot (#123)
golemfactory/golem-architecture: arch-snapshot/arch.md

> # Golem — current architecture

Reviewer: My main problem with this document is that it is sadly a tedious, uninspiring read. 

An architecture document should convey the most interesting and novel ideas as well and strive to inspire. It may be technically challenging but this should be balanced by a convinvcing story and gripping storytelling. Alas, this document fails to achieve either of this goals.

Authors: Witold Dzięcioł, Kamil Koczurek, Przemysław Rekucki, Marek Dopiera  
Reviewers: Maciej Maciejowski, Paweł Burgchardt  
Status: Review

# About this document

> The goal of this document is to describe present Golem architecture in enough

Reviewer: ```suggestion
The goal of this document is to describe the current Golem architecture in enough
```

detail for an outside person to understand how it works under the hood. The

> intended audience is assumed to be technical but not necessarily have deep
> expertise in the crypto world. The level of detail stops short of describing

Reviewer: ```suggestion
intended audience is assumed to be technical but not necessarily experts in the crypto world. The level of detail stops short of describing
```

code and its organization but describes key technologies used in order to
implement the needed functionalities.

The aim is not to dive deep into every integration, but to capture architectural
decisions and their consequences.

# General Concept

This section outlines the components of the Golem Network, including the actors,
technical artifacts, and the activities these actors may perform. The goal is to

> provide a high-level overview that ties together all key terms, offering a clear

Reviewer: ```suggestion
provide a high-level overview tying together all key terms, offering a clear
```

understanding of the network’s ecosystem.

## What Golem Network Is

Golem Network is an open-source, decentralized platform for trading and
utilizing computing power.

The following comparisons help build an intuitive understanding of Golem
Network. While each of them has its limitations, together they provide a first
approximation of what the Golem Network is.

### A Decentralized Compute Cloud

Golem Network can be compared to traditional cloud platforms like AWS or GCP, as
it enables users to run workloads on machines owned by others. However, there
are key differences:

* AWS and GCP own the infrastructure, whereas Golem Network relies on users
  renting out their own machines to others.
* Amazon and Google act as central authorities, but Golem gives no additional
  power to any single entity (including itself). Users can fork, modify, or
  build on top of the platform.
* Golem is not a service provider; it simply provides the tools for sellers and
  buyers to find each other and exchange resources for value.
* The range of predefined "services" on Golem is far smaller compared to

>   traditional cloud providers.

Reviewer: Maybe add something that this range keeps growing, provide examples of recent additions.

simplification, it is sufficient to grasp the core abstractions introduced by
the Golem Network.

### Provider

A **Provider** is a seller, someone who rents out their computing resources.
This could be as simple as renting out a personal laptop, or as extensive as
offering the resources of an entire data center.

The Provider sets the price for their resources, selects preferred payment
methods, and configures the necessary software, either provided by or built on
top of the Golem Network, to run on their machines. Once set up, these resources
can be utilized by others, and the Provider is compensated for their usage.

### Requestor

A **Requestor** is buyer, someone who wants to borrow computing resources. They
might do this to experiment with hardware they don’t own, to speed up
computations that would take too long on a single machine, or even to build a
service powered by resources from multiple Providers.

The Requestor needs to have a certain level of technical knowledge to interact
with Golem Network's APIs (e.g., using the Python or Node.js SDKs). Through
these APIs, they specify what resources they require and how they intend to pay

> for them.

Reviewer: This is not necessarily true; a Requestor (i.e. the entity demanding services and paying for them) may also have someone else write them the software needed.

> ## Activities

Reviewer: "Activity" has a specific meaning in Golem, so this section title should probably be changed

```suggestion
## Using the Golem network
```

> ## Activities

Reviewer: One major problem I have with this whole section is that it is provider-centriic, as if we did not care so much for requestors. I wonder if there may be a deeper issue underlying this approach. For example, compare the detailed description of "selling on Golem" vs the rudimentary subsection on buying.

This section describes what actors can do to the system. The descriptions are
only as detailed as to explain how the actors control the artifacts. The goal is

> to give the reader an overview of the terms introduced by Golem without any
> details and establish a glossary to ensure consistency within the document.

Reviewer: ```suggestion
to give the reader an overview of the terms introduced by Golem and establish a glossary to ensure consistency within the document.
```

> ### Selling on Golem platform

Reviewer: This section together with the "Buying on Golem" section give a distorted view of what is actually happening. In this description, a provider receives multiple proposals and chooses the best one, while the requestor accepts first offer matching their demand.

The reality is different (and has been for quite a while): providers wait long for a proposal and usually accept the first one that comes along, while requestors can and should choose the best offer.

When a Requestor finds a suitable Offer, they contact the Provider directly to negotiate the deal.

Typically, humans are not involved in the process of finding, matching, negotiating, or finalizing agreements. Instead,
users define their needs programmatically, allowing the [Provider Agent](#provider-agent) and [Requestor Agent](#requester-agent)
software to handle these tasks automatically.

The [Provider Agent](#provider-agent) is primarily responsible for implementing the logic needed to sell resources on
the Golem Network. From high level perspective, Provider Agent application should do following things: 
1. Describe Resources using property language to create an [Offer](#offer)
2. Publish the Offer on the market
3. Monitor incoming Proposals and negotiate an [Agreement](#agreement) with the most promising [Requestor](#Requestor)
4. Allocate the promised Resources in accordance with the [Agreement](#agreement)
5. Monitor resources usage and charge Requestor Agent
6. Terminate the Agreement or await the Agreement termination event from the [Requestor](#Requestor)
7. Send an [Invoice](#invoice) summarizing the total cost of the [Agreement](#agreement)

> 8. Wait until the payment for the [Invoice](#invoice) is settled and payment confirmed. 

Reviewer: This description may seem intimidating. It might be worth stating that a Golem user does not necessarily need to create a program to do all this things, but simply use a default Provider Agent with minimal configuration effort.

#### 1. Describe Resources using property language to create an [Offer](#offer)

While Golem is currently used for trading computational resources, it was designed to support the exchange of any type of

> [resource](#resource). This means the Marketplace does not enforce strict standards on the goods being traded.

Reviewer: ```suggestion
[resource](#resource). This means the Marketplace does not enforce strict standards on the services being traded.
```

  including the number of CPU cores, RAM, and disk space.
- The price and the [pricing model](#payments-models) applied.
- The payment scheme, outlining how the agent application interacts with the payment system (e.g. in `pay-as-you-use` scheme,
  what is expected transaction frequency).
- The [Wallet](#wallet) address for receiving payments, along with the supported [payment platforms](#payment-platform).

#### 2. Publish the Offer on the market

Golem is a decentralized network of independent [Nodes](#yagna-node), with no central repository for [Offers](#offer) or any
central server to facilitate [Agreements](#agreement) between parties. As a result, offers must be propagated between nodes,
and transactions are conducted through direct communication.

Developers don’t need to worry about [offer propagation](#offer-propagation). The responsibility for propagating offers
lies with the Core Network. The only task for the Provider Agent is to publish the offer on the market and listen
for incoming [Proposals](#proposal).

#### 3. Monitor incoming Proposals and negotiate an Agreement with the most promising Requestor

> The [Provider Agent](#provider-agent) plays a passive role in negotiations. Offers are propagated across the network and

Reviewer: Yet the diagram below states that the provider "updates Proposals" and sends "Counter Proposals". Is this passive?

> received by [Requestors](#Requestor). The offer is matched locally on the Requestor's node with a [Demand](#demand).

Reviewer: This is the first place the notion of Demand occurs. It might be worthwhile to explain what it is (especially that the linked definition is not very enlightening)

The negotiation stage serves several purposes:
- Ensures that the Provider Agent and Requestor Agent communicate before signing an Agreement (since offer propagation
  doesn’t require direct interaction between parties).
- Allows both the Provider Agent and Requestor Agent to implement different [strategies](#market-strategies) to maximize
  their benefits and select suitable partners.
- Provides an opportunity for the Provider Agent and Requestor Agent to negotiate additional terms that weren’t included
  in the initial [Proposals](#proposal). This is possible through protocols built on top of the property language.

Both the Provider Agent and Requestor Agent negotiate with multiple Agents simultaneously. The Requestor Agent initiates
the Agreement by proposing it to the Provider Agent, who can either accept or reject the Proposal. Once the Agreement is signed
by both parties, the Requestor Agent can begin using the resources. The Agreement remains valid until it is terminated by
either party. The terms of termination (e.g., duration of the Agreement and conditions under which it can be terminated) are
specified within the Agreement.

```mermaid
---

> title: Simplified negotiations from Provider Agent's perspective

Reviewer: This diagram should be made clearer

platform it supports. It is the Requestor Agent's responsibility to choose the platform by specifying the appropriate
[property](#property) in their demand. The Requestor Agent can approach negotiations in two ways:

###### 1. Static Negotiations
Suppose the Requestor Agent prefers payments on the Polygon network. In this case, they require the Provider Agent to support
Polygon and will not select a Provider Agent that doesn’t.

Since the Requestor has a specific requirement, multiple negotiation stages aren't necessary. They can simply add
a [constraint](#constraint) to their [Demand](#demand), instructing the matching algorithm to filter out Providers that
don’t meet this requirement. In their Demand, they set the chosen platform as a fixed value.

###### 2. Dynamic Negotiations
Now imagine a Requestor Agent that can pay on multiple platforms but prioritizes them based on transaction fees. In this
scenario, the Requestor Agent has a larger pool of potential Providers since they don’t restrict the platform by adding
a constraint to their demand.

Instead, the Requestor Agent collects Proposals from the market and evaluates them based on estimated costs. In later stages

> of Proposal exchange, they choose the platform by setting the relevant [property](#property) according to the Providers' scores,

Reviewer: Setting the property where? In the counter-proposal? But arch.md#property states "Property is a key-value pair present in Offers and Demands specifying properties of the Node producing the document.". This is confusing.

which are based on potential [transaction](#transaction-on-blockchain) costs.

> 

Reviewer: This negotiation example does not end with an agreement (in fact the outcome is not clear at all)

#### 4. Allocate the promised Resources in accordance with the Agreement

Once the [Agreement](#agreement) is signed, the Provider Agent is expected to reserve the promised [resources](#resource) for the

> Requestor’s use. During this time, the Provider Agent cannot sell these resources to anyone else and must be prepared

Reviewer: Reserve for how long? Indefinitely? What prevents an evil requestor from blocking provider resources without starting an Activity (and hence without payment)?

Invoices are issued after the Agreement is terminated, providing a summary of the total costs. They allow the Provider
Agent to include any additional costs not covered in the Debit Notes, as the final Debit Note doesn’t have to be sent
immediately after the activity ends.

```mermaid
flowchart LR
Activity1((Activity 1)) --o D11[Debit Note 1] --> D12[Debit Note 2] -->|...| D13[Debit Note N-th] --> Invoice[Invoice]
Activity2((Activity 2)) --o D21[Debit Note 1] --> D22[Debit Note 2] -->|...| D23[Debit Note N-th] --> Invoice[Invoice]
Activity3((Activity 3)) --o D31[Debit Note 1] --> D32[Debit Note 2] -->|...| D33[Debit Note N-th] --> Invoice[Invoice]
```

Both Debit Notes and Invoices can be either accepted or rejected by the other party. Acceptance signals that the
Requestor Agent agrees to pay the specified amount. Rejection, on the other hand, indicates refusal to pay the
non-accepted amount. However, it’s important to note that a rejection does not absolve the Requestor Agent from paying
for all previously accepted Debit Notes. The conditions under which rejection is allowed should be defined in the
Agreement. Currently, no payment scheme permits rejections.

> 

Reviewer: How can a Requestor safeguard against a dishonest provider?
Or even honest mistakes, if "no payment scheme permits rejections"?

Debit Notes can be classified as payable or non-payable, with payable Debit Notes identified by the due date included
in the document. While payable Debit Notes are scheduled for processing upon acceptance, this still does not necessitate
immediate payment. The payment mechanism allows for the [batching of payments](#payments-batching) or delaying them
to accommodate additional Debit Notes or [Invoices](#invoice), thereby reducing [transaction](#transaction-on-blockchain) costs on the blockchain.

The consequence of delaying payments is that they are not guaranteed. However, this design opens the possibility of
implementing mechanisms that can mitigate or eliminate the risk of non-payment. For instance, a payment platform
could be developed using a deposit or escrow contract, or by integrating payment channels into the Core Network.

It’s important to note that, regardless of the payment scheme or platform used, Golem Factory does not act as an
intermediary for payments. Since transactions occur on the blockchain, and due to the decentralized nature of blockchain
technology, Golem Factory has no control over these transactions.

> 

Reviewer: The text from here to end of section is repeated instance of the text above.

Accepting a Debit Note or Invoice does not result in immediate payment for a few reasons.
Debit Notes can be classified as payable or non-payable, with payable Debit Notes identified by the due date included
in the document. While payable Debit Notes are scheduled for processing upon acceptance, this still does not necessitate
immediate payment. The payment mechanism allows for the [batching of payments](#payments-batching) or delaying them
to accommodate additional Debit Notes or [Invoices](#invoice), thereby reducing [transaction](#transaction-on-blockchain) costs on the blockchain.

The consequence of delaying payments is that they are not guaranteed. However, this design opens the possibility of
implementing mechanisms that can mitigate or eliminate the risk of non-payment. For instance, a payment platform
could be developed using a deposit or escrow contract, or by integrating payment channels into the Core Network.

It’s important to note that, regardless of the payment scheme or platform used, Golem Factory does not act as an
intermediary for payments. Since transactions occur on the blockchain, and due to the decentralized nature of blockchain

> technology, Golem Factory has no control over these transactions.

Reviewer: duplicated text

permitted. Below is a non-exhaustive list of possible reasons for termination:
- The Agreement expires if it was established for a fixed duration.
- The Requestor Agent no longer needs the [resources](#resource) or has completed the computations.
- One of the parties violates the terms of the Agreement, such as:
  - The Requestor Agent fails to accept [Debit Notes](#debit-note) within the agreed timeframe.
  - The Provider Agent issues Debit Notes more frequently than agreed.
  - The Requestor Agent fails to make timely payments, particularly in cases involving [mid-agreement payments](#mid-agreement-payments).

It is the Agent—whether Requestor or Provider—who decides to terminate the Agreement. The Agent is also responsible for
detecting if the other party has terminated the Agreement and taking the appropriate action in response.

Provider Agent has the option to attach additional information outlining the reasons for termination when ending the
Agreement. While this is not mandatory, it is encouraged as it can provide valuable context for the other party,
serving as diagnostic information or for other purposes.

> The [Agreement](#agreement) can be terminated when either party chooses to end it. Core Network doesn't enforce any

Reviewer 2: Duplication with line 389 - 404

Agreement. While this is not mandatory, it is encouraged as it can provide valuable context for the other party,
serving as diagnostic information or for other purposes.

#### 7. Send an Invoice summarizing the total cost of the Agreement

Once the Agreement is terminated, the Provider Agent should send an [Invoice](#invoice) to the Requestor Agent summarizing
the total costs incurred throughout the Agreement. This Invoice should reflect the cumulative costs from all [Activities](#activity).
In response, the Requestor Agent must either accept or reject the Invoice. However, regardless of the acceptance status,
payment is mandatory for the total amount indicated by the accepted [Debit Notes](#debit-note), as their acceptance constitutes
a binding commitment to pay.

```mermaid
flowchart LR
Activity1((Activity 1)) --o D11[Debit Note 1] --> D12[Debit Note 2] -->|...| D13[Debit Note N-th] --> Invoice[Invoice]
Activity2((Activity 2)) --o D21[Debit Note 1] --> D22[Debit Note 2] -->|...| D23[Debit Note N-th] --> Invoice[Invoice]
Activity3((Activity 3)) --o D31[Debit Note 1] --> D32[Debit Note 2] -->|...| D33[Debit Note N-th] --> Invoice[Invoice]
```

> Once the Agreement is terminated, the Provider Agent should send an [Invoice](#invoice) to the Requestor Agent summarizing

Reviewer 2: Duplication of lines 425 - 429

the total costs incurred throughout the Agreement. This Invoice should reflect the cumulative costs from all [Activities](#activity).
In response, the Requestor Agent must either accept or reject the Invoice. However, regardless of the acceptance status,
payment is mandatory for the total amount indicated by the accepted [Debit Notes](#debit-note), as their acceptance constitutes
a binding commitment to pay.

#### 8. Wait until the payment for the Invoice is settled and payment confirmed.

It's important for the Provider to monitor payments after the Agreement is completed. This is when the Provider Agent
should adjust its market strategy to ensure profitability. Since the Core Network doesn't guarantee payment delivery,
the Provider Agent should implement measures to prevent being exploited by Requestors. One example is rejecting

> non-paying Requestors and prioritizing those with a good reputation. Lack of payment isn't the only reason for

Reviewer 2: `reputation` -> Here, it would be probably good to explain what reputation is. Right now it’s unclear, and the protocol does not support something like Reputation... but from this description someone might assume that reputation is built into the protocol.

declining a Requestor in the future. The Provider Agent may also choose to reject subsequent Agreements with
Requestors who break the Agreement conditions.

Payment confirmation is received by the Provider Agent from the Requestor once the transaction is confirmed on the
blockchain. This confirmation specifies which Activities and Agreements are covered by the transaction. There is no
1-to-1 relationship between transactions and Activities or Agreements. A single blockchain transaction can cover
multiple Activities or Agreements, while each Activity or Agreement may also be covered by multiple transactions.

Payments are not immediate for several reasons: they are not scheduled right away, and batching 
may occur. Furthermore, blockchain transactions are not immediate and may take time to process. Therefore, the Provider
Agent should monitor Payment events. This can be done by listening for status changes to Settled on Invoice and Debit Note
events, or by tracking payment events to receive notifications for each transaction.

> Payment confirmation is received by the Provider Agent from the Requestor once the transaction is confirmed on the

Reviewer 2: Duplication with 454 - 457

1-to-1 relationship between transactions and Activities or Agreements. A single blockchain transaction can cover
multiple Activities or Agreements, while each Activity or Agreement may also be covered by multiple transactions.

### Searching on market

Providers or Requestors on Golem don’t always aim to buy or sell resources. Sometimes, they simply want to observe the market.
There are several possible reasons for this:
- Checking which resources are available on the network
- Estimating potential computation costs
- Gathering insights to better adjust negotiation strategies
- Collecting statistics about the network

The approach to observing the market is similar to the methods described in the chapters on buying and selling resources.
An agent specifies their Offer or Demand using a property and constraints specification language.
The key difference is that they are not required to negotiate with other agents, nor must they provide all necessary
information in their Offer or Demand. This is a more lightweight approach compared to going through the full negotiation
process.

> ### Buying on golem platform.

Reviewer: ```suggestion
### Buying on Golem platform.
```

Buying on Golem Network, just like [selling](#selling-on-golem-platform),
involves the [Requestor](#requestor) specifying what one needs in a formal

> language, letting the platform match Offers and a [Requestor

Reviewer 2: What does it mean that the platform matches offers? From earlier descriptions, it seems there is no automatic offer matching, only propagation. It is the responsibility of the respective agents to match offers, not the platform.

Agent](#requester-agent) software negotiate and seal the deal.

More specifically, the the Requestor and Requestor Agent need to:
1. Configure and fund the Wallet
1. Allocate funds
1. Create a Demand
1. Negotiate and Agreement
1. Activate the service and supervise the Agreement 

> 1. Closure the Agreement

Reviewer: I think "closure" is usually used as a noun rather than a verb.

#### 1. Configure and fund the Wallet

Before the Requestor begins, they must secure appropriate funds. To do this,
they should have funds on the wallet address from which they will pay, on one of

> the two supported blockchains: Ethereum or Polygon. It is strongly recommended

Reviewer 2: This is the case now, but shouldn't it be mentioned that these platforms are more generic than just two?

The Requestor (a person) does not browse through all Providers' Offers on their
own. Instead, the Requestor Agent is used to programmatically specify which
Offers are acceptable. Such a specification is called a Demand in the Golem
nomenclature. The Requestor Agent doesn't send a Demand anywhere - it is its
internal, ephemeral way of filtering the Offers.

A requestor can specify the same things in a Demand, which the Provider's can in
an Offer. This was described in detail in the [section about
selling](#1-describe-resources-using-property-language-to-create-an-offer).

#### 4. Negotiate and Agreement

As Offers are propagated through Golem Network, the matching ones are delivered
to the Requestor Agent. The section on [Selling on
Golem](#3-monitor-incoming-proposals-and-negotiate-an-agreement-with-the-most-promising-requestor)
describes this process in detail.

#### 5. Activate the service and supervise the Agreement 

With an active Agreement, the Requestor activates the service by creating an
Activity. For a more detailed description defer to the section on [Selling on
Golem](#5-monitor-resources-usage-and-charge-requestor-agent).

#### 6. Closure the Agreement

> The Requestor is responsible for terminating the Agreement.  This notifies the

Reviewer 2: We previously wrote that both parties can terminate the agreement. I
s it okay to indicate the Requestor as the responsible party here?

Provider to terminate all activities associated with that contract.  The
Requestor receives an Invoice summarizing the expenses from all Activities

> active under the agreement.  The Requestor confirms that the amount is correct.

Reviewer: and if it is not correct?

After some time, Golem transfers tokens to Provider's wallet as payment for the
service.

## Layers

> Golem Network's applications are unknown in advance. Therefore, based on past

Reviewer 2: A very strange introduction for the beginning of a section on architecture layers

experience, a decision was made to design the architecture flexible enough to

> swap out key components. This resulted in a three-layer architecture.

Reviewer: It might be helpful to explain how the layers communicate, since from this descriptions they seem to be mostly independent from each other, so as to represent "dimensions" rather than "layers"

   where communication is relayed, there is no Session-establishing phase between parties sending relayed packets, 
   so key exchanges aren't feasible.

For these reasons, the AES-GCM-SIV variant of the AES algorithm was chosen.

**Symmetric keys derivation**

Hybrid Net employs Elliptic Curve Diffie-Hellman (ECDH) to derive symmetric encryption keys. Each Node generates a 
new private/public key pair, separate from its identity keys, which is specifically used for communication. The 
public key is exchanged during the [p2p handshake](#establishing-sessions-between-nodes), and a shared secret is 
derived from this key to enable AES encryption.

In cases of relayed communication, where no direct Session is established, the public key is retrieved from the 
Relay server. The shared secret is then derived using the Node's private key, enabling encrypted communication 
through the Relay. The recipient Node also requests the sender's public key from the Relay server and uses it to 
derive the shared secret, allowing it to decrypt the message. No special control packet exchanges are required 
between the sender and receiver.

> #### Central net

Reviewer: Description missing

### Market interactions
A description of the component responsible for making offers, counter-offers,

> negotiations, etc.

Reviewer: description missing

There is no single, comprehensive document listing all properties and protocols. However, specifications are stored
in several places:
- Some features are initially proposed through [Golem Amendment Proposals](../gaps/) (GAPs) and then implemented.
- Other features are documented after implementation in the [specs](../specs/) directory.
- Certain properties not tied to specific features can be found in the [standards](../standards) directory and the  
  corresponding entry in [cheat sheet](../standards/cheat_sheet.md).

Since the protocol is open to extensions and contributions from outside Golem Factory, certain issues may arise.
Introducing new properties could conflict with existing protocols, leading to naming clashes when multiple parties
introduce them simultaneously.

Currently, Golem has no formal strategy for managing these conflicts. Although solutions such as
[experimental features](../gaps/gap-32_experimental_features) have been proposed (mainly for internal purposes), a
more comprehensive process for managing protocols would be necessary if Golem reaches full decentralization. However,
at this moment, the problem remains non-existent.

#### Offer propagation

> - Link to design [decision](#only-providers-offers-are-propagated)

Reviewer: broken link

    
    par
      RequestorYagna->>ProviderYagna: Counter Proposal
      ProviderYagna->>ProviderAgent: Receive Proposal
    and Proposals from other Nodes in the network
      GolemNetwork->>ProviderAgent: Receive Proposals    
    end

    ProviderAgent->>ProviderAgent: Select best Proposals to respond
    ProviderAgent->>ProviderAgent: Adjust Proposals
    ProviderAgent->>ProviderYagna: Counter Proposal
    ProviderYagna->>RequestorYagna: Counter Proposal
    RequestorYagna->>RequestorAgent: Receive Proposal
  end

  RequestorAgent->>RequestorYagna: Propose Agreement
  RequestorYagna->>ProviderYagna: Propose Agreement
  ProviderYagna->>ProviderAgent: Receive Agreement Proposal
  ProviderAgent->>ProviderAgent: Select best Agreement Proposal
  ProviderAgent->>ProviderYagna: Approve Agreement
  ProviderYagna->>RequestorYagna: Approve Agreement Proposal
  RequestorYagna->>RequestorAgent: Agreement approval notification

  ProviderAgent->>ProviderYagna: Unsubscribe Proposal events
  deactivate ProviderYagna
  ProviderYagna-->GolemNetwork: Stop Offer propagation
  deactivate GolemNetwork
  deactivate RequestorYagna
```

#### Market strategies

> 

Reviewer: Empty section

#### Agreement termination
- Who is allowed to terminate? In what situation?
- What is specified by protocol and what is left to future specifications?

> - Termination reason concept

Reviewer: TODO
